### PR TITLE
Disentangle Nested If Else in Select Region and Add Region Class Selections

### PR DESCRIPTION
**Pull request by ctuguinay**
This addresses #131 and #140

### CODE REVIEW

File: echoregions/regions2d/regions2d.py [hunk 1]
```diff
@@ -107,132 +110,128 @@ def select_region(
...
-                        else:
-                            raise ValueError(
-                                f"1st index value must be later than 0th index \
-                                             value. Currently 0th index value is {time_range[0]} \
-                                             and 1st index value is {time_range[1]}"
-                            )
-                    else:
+            elif isinstance(region_id, list):
+                for value in region_id:
+                    if not isinstance(value, (float, int, str)):
```
Comment (leewujung) [inline echoregions/regions2d/regions2d.py:122]: seems like the list comprehension you have below for the timestamp part looks more succinct?
```
all(isinstance(t, Timestamp) for t in time_range)
```

File: echoregions/regions2d/regions2d.py [hunk 2]
```diff
@@ -107,132 +110,128 @@ def select_region(
...
-                    else:
+            region = region[region["region_id"].isin(region_id)]
+
+        # Check and subset for region_class
+        if region_class is not None:
+            if isinstance(region_class, str):
+                region_class = [region_class]
+            elif isinstance(region_class, list):
+                for value in region_class:
+                    if not isinstance(value, str):
```
Comment (leewujung) [inline echoregions/regions2d/regions2d.py:140]: same as the `region_id` comment: seems like the list comprehension you have below for the timestamp part looks more succinct?
```
all(isinstance(t, Timestamp) for t in time_range)
```

**Review by leewujung — CHANGES_REQUESTED** [review]
Hey @ctuguinay : Thanks for the PR! I made an inline comment for a minor style thing, and below are a couple things to take care of before we merge this:
- The "Returns" part of the docstring does not current include description for the case when `region_class` is given
- Right now there is no check for the case when `region_id` and `region_class` both exist, and the code would break if both are given as inputs. How about adding a check to raise an error on the case where both exist and the case where both are `None`? i.e., we require only one of them given as input. The docstring should say something about this too.

I was going to make a comment about the `select_sonar_file` function, but think that we should just leave it for another PR that addresses #126 -- that the file selection is based on timestamp for generic `xr.Dataset` objects.

File: echoregions/regions2d/regions2d.py (hunk 1 above)
Comment (leewujung) [reply echoregions/regions2d/regions2d.py:122]: Oh actually I see now that you use a for loop to print the exact type out. I'll resolve this comment.

File: echoregions/regions2d/regions2d.py [hunk 3]
```diff
@@ -107,132 +110,128 @@ def select_region(
...
-                    else:
+            elif isinstance(region_id, list):
+                for value in region_id:
+                    if not isinstance(value, (float, int, str)):
                         raise TypeError(
-                            f"Invalid time_range value types: \
-                                        {type(time_range[0])} and {type(time_range[1])}. Must \
-                                        be both of type Timestamp."
+                            f"Invalid element in list region_id. Is of type: "
+                            f"{type(value)}. Must be of type float, int, str."
```
Comment (leewujung) [inline echoregions/regions2d/regions2d.py:125]:
```suggestion
                            f"Invalid element in list region_id of type: "
                            f"{type(value)}. Must be of type float, int, str."
```

File: echoregions/regions2d/regions2d.py (hunk 2 above)
Comment (leewujung) [reply echoregions/regions2d/regions2d.py:140]: Same as the above, I'll resolve this now.

File: echoregions/regions2d/regions2d.py [hunk 4]
```diff
@@ -107,132 +110,128 @@ def select_region(
...
+                region_class = [region_class]
+            elif isinstance(region_class, list):
+                for value in region_class:
+                    if not isinstance(value, str):
                         raise TypeError(
-                            f"Invalid depth_range value types: \
-                                        {type(depth_range[0])} and {type(depth_range[1])}. Must \
-                                        be both of type either float or int."
+                            f"Invalid element in list region_class. Is of type: "
+                            f"{type(value)}. Must be of type str."
```
Comment (leewujung) [inline echoregions/regions2d/regions2d.py:143]:
```suggestion
                            f"Invalid element in list region_class of type: "
                            f"{type(value)}. Must be of type str."
```

**Comment by ctuguinay** [comment]
@leewujung Thanks for the comments! I just have a few questions before I start working on this again.

> Right now there is no check for the case when region_id and region_class both exist, and the code would break if both are given as inputs. How about adding a check to raise an error on the case where both exist and the case where both are None? i.e., we require only one of them given as input. The docstring should say something about this too.

Why exactly would the code break for both `region_id` and `region_class`? The operations for checking and subsetting `region_id` and `region_class` are independent of one another https://github.com/OSOceanAcoustics/echoregions/blob/83781bcc575620562faf3391bb1d39bd309142af/echoregions/regions2d/regions2d.py#L116C1-L150C71.

I also think being able to combine both of these as selection parameters should still be the desired behavior. I can't see a reason why we, in particular, would need both `region_id` and `region_class` to be non-NaN, but I don't necessarily think that this type of selection should be restricted since we do allow the other selections to be combined, i.e. `region_id`, `depth_range`, and `time_range`.

**Comment by leewujung** [comment]
> I also think being able to combine both of these as selection parameters should still be the desired behavior. I can't see a reason why we, in particular, would need both `region_id` and `region_class` to be non-NaN, but I don't necessarily think that this type of selection should be restricted since we do allow the other selections to be combined, i.e. `region_id`, `depth_range`, and `time_range`.

My rationale for restricting the selection to be only `region_id` and `region_class` is just to avoid confusion and user error in the form of unintended additional filtering of the dataframe. I think that if people want to select only some `region_id` that are within a particular `region_class`, they really should only pass the specific `region_id` they want (i.e. doing the selection outside of this function), instead of passing in the list of `region_id` and expect `region_class` to take care of the additional
filtering to remove IDs that are not in a particular `region_class`.

Also, since filtering in the code section you highlighted are sequential, they are not independent -- `region_class` here is actually an additional filtering criteria after `self.data` has been filtered with the `region_id` input argument.

**Comment by ctuguinay** [comment]
@leewujung I added all your previous suggestions. This is ready to be reviewed again.

**Review by leewujung — APPROVED** [review]
@ctuguinay : All looks good! I pushed some small tweaks of `select_region`. Once you verify those, feel free to merge!

**Pull request merged**